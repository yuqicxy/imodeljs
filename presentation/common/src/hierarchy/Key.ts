/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { InstanceKey, InstanceKeyJSON } from "../EC";

/**
 * Standard node types
 * @public
 */
export enum StandardNodeTypes {
  /** @deprecated Use `ECInstancesNode` */
  ECInstanceNode = "ECInstanceNode",
  ECInstancesNode = "ECInstancesNode",
  ECClassGroupingNode = "ECClassGroupingNode",
  ECPropertyGroupingNode = "ECPropertyGroupingNode",
  DisplayLabelGroupingNode = "DisplayLabelGroupingNode",
}

/**
 * One of the node key types
 * @public
 */
export type NodeKey = BaseNodeKey | ECInstanceNodeKey | ECInstancesNodeKey | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
/** @public */
export namespace NodeKey {
  /**
   * Serialize given [[NodeKey]] to JSON
   * @internal
   */
  export function toJSON(key: NodeKey): NodeKeyJSON {
    if (isInstancesNodeKey(key))
      return { ...key, instanceKeys: key.instanceKeys.map(InstanceKey.toJSON) };
    if (isInstanceNodeKey(key))
      return { ...key, instanceKey: InstanceKey.toJSON(key.instanceKey) };
    return { ...key };
  }

  /**
   * Deserialize node key from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized node key
   *
   * @internal
   */
  export function fromJSON(json: NodeKeyJSON): NodeKey {
    if (isInstancesNodeKey(json))
      return { ...json, instanceKeys: json.instanceKeys.map(InstanceKey.fromJSON), instanceKey: InstanceKey.fromJSON(json.instanceKeys[0]) };
    if (isInstanceNodeKey(json))
      return { ...json, instanceKey: InstanceKey.fromJSON(json.instanceKey) };
    return { ...json };
  }

  /**
   * Checks if the supplied key is an [[ECInstanceNodeKey]]
   * @public
   * @deprecated Will be removed with `StandardNodeTypes.ECInstanceNode`. Consider using `StandardNodeTypes.ECInstancesNode` and related APIs.
   */
  export function isInstanceNodeKey(key: NodeKey): key is ECInstanceNodeKey {
    return key.type === StandardNodeTypes.ECInstanceNode || isInstancesNodeKey(key);
  }

  /**
   * Checks if the supplied key is an [[ECInstancesNodeKey]]
   * @public
   */
  export function isInstancesNodeKey(key: NodeKey): key is ECInstancesNodeKey {
    return key.type === StandardNodeTypes.ECInstancesNode;
  }

  /**
   * Checks if the supplied key is an [[ECClassGroupingNodeKey]]
   * @public
   */
  export function isClassGroupingNodeKey(key: NodeKey): key is ECClassGroupingNodeKey {
    return key.type === StandardNodeTypes.ECClassGroupingNode;
  }

  /**
   * Checks if the supplied key is an [[ECPropertyGroupingNodeKey]]
   * @public
   */
  export function isPropertyGroupingNodeKey(key: NodeKey): key is ECPropertyGroupingNodeKey {
    return key.type === StandardNodeTypes.ECPropertyGroupingNode;
  }

  /**
   * Checks if the supplied key is a [[LabelGroupingNodeKey]]
   * @public
   */
  export function isLabelGroupingNodeKey(key: NodeKey): key is LabelGroupingNodeKey {
    return key.type === StandardNodeTypes.DisplayLabelGroupingNode;
  }

  /**
   * Checks if the supplied key is a grouping node key
   * @public
   */
  export function isGroupingNodeKey(key: NodeKey): key is GroupingNodeKey {
    return isClassGroupingNodeKey(key) || isPropertyGroupingNodeKey(key) || isLabelGroupingNodeKey(key);
  }
}

/**
 * Node key path. Can be used to define path from one node to another.
 * @public
 */
export type NodeKeyPath = NodeKey[];

/**
 * Data structure that describes a basic node key
 * @public
 */
export interface BaseNodeKey {
  /** Node type */
  type: string;
  /** Node hash path from root to the node whose key this is */
  pathFromRoot: string[];
}

/**
 * Data structure that describes an ECInstance node key
 * @public
 * @deprecated Use `ECInstancesNodeKey`
 */
export interface ECInstanceNodeKey extends BaseNodeKey {
  type: StandardNodeTypes.ECInstanceNode;
  /** [[InstanceKey]] of the ECInstance represented by the node */
  instanceKey: InstanceKey;
}

/**
 * Serialized [[ECInstanceNodeKey]] JSON representation.
 * @internal
 * @deprecated Use `ECInstancesNodeKeyJSON`
 */
export interface ECInstanceNodeKeyJSON extends BaseNodeKey {
  type: StandardNodeTypes.ECInstanceNode;
  instanceKey: InstanceKeyJSON;
}

/**
 * Data structure that describes a node ECInstance node key
 * @public
 */
export interface ECInstancesNodeKey extends BaseNodeKey {
  type: StandardNodeTypes.ECInstancesNode;
  /** List of [[InstanceKey]] objects of ECInstances represented by the node */
  instanceKeys: InstanceKey[];
  /**
   * One of the instance keys from [[instanceKeys]] array. Here only to avoid breaking
   * consumers' code after switching from [[ECInstanceNodeKey]].
   *
   * @alpha Added temporarily until [[ECInstanceNodeKey]] is completely removed. Use [[instanceKeys]] instead.
   */
  instanceKey: InstanceKey;
}

/**
 * Serialized [[ECInstanceNodeKey]] JSON representation.
 * @internal
 */
export interface ECInstancesNodeKeyJSON extends BaseNodeKey {
  type: StandardNodeTypes.ECInstancesNode;
  instanceKeys: InstanceKeyJSON[];
}

/**
 * Data structure that describes a grouping node key
 * @public
 */
export interface GroupingNodeKey extends BaseNodeKey {
  /**
   * Get the number of instances grouped by the node represented
   * by this key.
   *
   * **Note:** this property is just a helper and is not involved
   * in identifying a node.
   */
  groupedInstancesCount: number;
}

/**
 * Data structure that describes an ECClass grouping node key
 * @public
 */
export interface ECClassGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECClassGroupingNode;
  /** Full name of the grouping ECClass */
  className: string;
}

/**
 * Data structure that describes an ECProperty grouping node key
 * @public
 */
export interface ECPropertyGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.ECPropertyGroupingNode;
  /** Full name of the grouping ECProperty class */
  className: string;
  /** Name of the ECProperty */
  propertyName: string;
  /** Grouping value */
  groupingValue: any;
}

/**
 * Data structure that describes a display label grouping node key
 * @public
 */
export interface LabelGroupingNodeKey extends GroupingNodeKey {
  type: StandardNodeTypes.DisplayLabelGroupingNode;
  /** Grouping display label */
  label: string;
}

/**
 * One of the serialized node key types
 * @internal
 */
export type NodeKeyJSON = BaseNodeKey | ECInstanceNodeKeyJSON | ECInstancesNodeKeyJSON | ECClassGroupingNodeKey | ECPropertyGroupingNodeKey | LabelGroupingNodeKey;
