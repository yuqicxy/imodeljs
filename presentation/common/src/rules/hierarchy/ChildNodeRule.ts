/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { NavigationRuleBase } from "./NavigationRule";
import { RuleTypes, ConditionContainer } from "../Rule";

/**
 * Child node rules define nodes that are displayed at
 * each child hierarchy level.
 *
 * @public
 */
export interface ChildNodeRule extends NavigationRuleBase, ConditionContainer {
  /** Used for serializing to JSON. */
  ruleType: RuleTypes.ChildNodes;

  /**
   * Defines a condition for the rule, which needs to be met in order to execute it. Condition
   * is an [ECExpression]($docs/learning/presentation/ECExpressions.md), which can use
   * a [limited set of symbols]($docs/learning/presentation/Hierarchies/ECExpressions.md#rule-condition).
   */
  condition?: string;
}
