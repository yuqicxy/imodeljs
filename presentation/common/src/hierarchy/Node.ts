/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { NodeKey, NodeKeyJSON } from "./Key";
import { LabelDefinitionJSON, LabelDefinition } from "../LabelDefinition";

/**
 * Data structure that describes a tree node.
 * @public
 */
export interface Node {
  /** A key that uniquely identifies a node. */
  key: NodeKey;
  /** Display label
   * @deprecated use 'labelDefinition' instead
   */
  label: string;
  /** Definition of node display label */
  labelDefinition?: LabelDefinition;
  /** Extensive description */
  description?: string;
  /** Image ID */
  imageId?: string;
  /** Foreground color */
  foreColor?: string;
  /** Background color */
  backColor?: string;
  /** Font style */
  fontStyle?: string;
  /** Does this node have child nodes */
  hasChildren?: boolean;
  /** Is this node selectable */
  isSelectionDisabled?: boolean;
  /** Is this node editable */
  isEditable?: boolean;
  /** Is this node expanded */
  isExpanded?: boolean;
  /** Is checkbox visible for this node */
  isCheckboxVisible?: boolean;
  /** Is this node's checkbox checked */
  isChecked?: boolean;
  /** Is this node's checkbox enabled */
  isCheckboxEnabled?: boolean;
  /** Extended data injected into this node */
  extendedData?: { [key: string]: any };
}

/**
 * Serialized [[Node]] JSON representation.
 * @internal
 */
export interface NodeJSON {
  key: NodeKeyJSON;
  /** @deprecated use labelDefinition instead */
  label?: string;
  labelDefinition: LabelDefinitionJSON;
  description?: string;
  imageId?: string;
  foreColor?: string;
  backColor?: string;
  fontStyle?: string;
  hasChildren?: boolean;
  isSelectionDisabled?: boolean;
  isEditable?: boolean;
  isExpanded?: boolean;
  isCheckboxVisible?: boolean;
  isChecked?: boolean;
  isCheckboxEnabled?: boolean;
  extendedData?: { [key: string]: any };
}
/** @public */
export namespace Node {
  /**
   * Serialize given node to JSON.
   * @internal
   */
  export function toJSON(node: Node): NodeJSON {
    return {
      ...node,
      key: NodeKey.toJSON(node.key),
      labelDefinition: LabelDefinition.toJSON(node.labelDefinition ? node.labelDefinition : LabelDefinition.fromLabelString(node.label)),
      label: (node.labelDefinition && node.labelDefinition.displayValue) || node.label,
    };
  }

  /**
   * Deserialize node from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized node
   *
   * @internal
   */
  export function fromJSON(json: NodeJSON | string): Node {
    if (typeof json === "string")
      return JSON.parse(json, reviver);
    return Object.assign({}, json, {
      key: NodeKey.fromJSON(json.key),
      labelDefinition: LabelDefinition.fromJSON(json.labelDefinition),
      label: json.labelDefinition.displayValue,
    });
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[Node]] objects.
   *
   * @internal
   */
  export function reviver(key: string, value: any): any {
    return key === "" ? fromJSON(value) : value;
  }

  /**
   * Deserialize nodes list from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized nodes list
   *
   * @internal
   */
  export function listFromJSON(json: NodeJSON[] | string): Node[] {
    if (typeof json === "string")
      return JSON.parse(json, listReviver);
    return json.map((m) => fromJSON(m));
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[Node]][] objects.
   *
   * @internal
   */
  export function listReviver(key: string, value: any): any {
    return key === "" ? listFromJSON(value) : value;
  }
}
