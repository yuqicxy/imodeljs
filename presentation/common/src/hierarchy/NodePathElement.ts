/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

import { Node, NodeJSON } from "./Node";

/**
 * Data related to node hierarchy filtering
 * @public
 */
export interface NodePathFilteringData {
  /** Number of filter matches in the current element */
  matchesCount: number;
  /** Number of filter matches in the current element's children (recursively) */
  childMatchesCount: number;
}

/**
 * Describes a single step in the nodes path.
 * @public
 */
export interface NodePathElement {
  /** Node instance */
  node: Node;
  /** Node index  */
  index: number;
  /** Is this element part of the marked path */
  isMarked?: boolean;
  /** Child path elements */
  children: NodePathElement[];
  /** Additional filtering-related information */
  filteringData?: NodePathFilteringData;
}
/** @public */
export namespace NodePathElement {
  /**
   * Serialize given [[NodePathElement]] to JSON
   * @internal
   */
  export function toJSON(npe: NodePathElement): NodePathElementJSON {
    const result: NodePathElementJSON = {
      node: Node.toJSON(npe.node),
      index: npe.index,
      children: npe.children.map(NodePathElement.toJSON),
    };
    if (undefined !== npe.isMarked)
      result.isMarked = npe.isMarked;
    if (undefined !== npe.filteringData)
      result.filteringData = nodePathFilteringDataToJson(npe.filteringData);
    return result;
  }

  /**
   * Deserialize [[NodePathElement]] from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized [[NodePathElement]]
   *
   * @internal
   */
  export function fromJSON(json: NodePathElementJSON | string): NodePathElement {
    if (typeof json === "string")
      return JSON.parse(json, reviver);
    const result: NodePathElement = {
      index: json.index,
      node: Node.fromJSON(json.node),
      children: listFromJSON(json.children),
    };
    if (undefined !== json.isMarked)
      result.isMarked = json.isMarked;
    if (undefined !== json.filteringData)
      result.filteringData = nodePathFilteringDataFromJson(json.filteringData);
    return result;
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[NodePathElement]] objects.
   *
   * @internal
   */
  export function reviver(key: string, value: any): any {
    return key === "" ? fromJSON(value) : value;
  }

  /**
   * Deserialize [[NodePathElement]] list from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized [[NodePathElement]] list
   *
   * @internal
   */
  export function listFromJSON(json: NodePathElementJSON[] | string): NodePathElement[] {
    if (typeof json === "string")
      return JSON.parse(json, listReviver);
    return json.map((m) => fromJSON(m));
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[NodePathElement]][] objects.
   *
   * @internal
   */
  export function listReviver(key: string, value: any): any {
    return key === "" ? listFromJSON(value) : value;
  }
}

/**
 * Serialized [[NodePathFilteringData]] JSON representation.
 * @internal
 */
export interface NodePathFilteringDataJSON {
  occurances: number;
  childrenOccurances: number;
}

/**
 * Serialized [[NodePathElement]] JSON representation.
 * @internal
 */
export interface NodePathElementJSON {
  node: NodeJSON;
  index: number;
  isMarked?: boolean;
  children: NodePathElementJSON[];
  filteringData?: NodePathFilteringDataJSON;
}

const nodePathFilteringDataToJson = (npfd: NodePathFilteringData): NodePathFilteringDataJSON => {
  return {
    occurances: npfd.matchesCount,
    childrenOccurances: npfd.childMatchesCount,
  };
};

const nodePathFilteringDataFromJson = (json: NodePathFilteringDataJSON): NodePathFilteringData => {
  return {
    matchesCount: json.occurances,
    childMatchesCount: json.childrenOccurances,
  };
};
