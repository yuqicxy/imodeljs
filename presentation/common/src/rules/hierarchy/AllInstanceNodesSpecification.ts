/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */
import { ChildNodeSpecificationBase, ChildNodeSpecificationTypes, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";
import { SchemasSpecification } from "../SchemasSpecification";

/**
 * Creates nodes for all available instances filtered only by the
 * [[supportedSchemas]] properties of the specification or the ruleset.
 *
 * @public
 */
export interface AllInstanceNodesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes.AllInstanceNodes;

  /**
   * Specification of schemas whose instances should be returned.
   */
  supportedSchemas?: SchemasSpecification;
}
