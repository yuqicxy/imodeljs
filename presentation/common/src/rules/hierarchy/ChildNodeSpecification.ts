/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { ChildNodeRule } from "./ChildNodeRule";
import { AllInstanceNodesSpecification } from "./AllInstanceNodesSpecification";
import { AllRelatedInstanceNodesSpecification } from "./AllRelatedInstanceNodesSpecification";
import { CustomNodeSpecification } from "./CustomNodeSpecification";
import { InstanceNodesOfSpecificClassesSpecification } from "./InstanceNodesOfSpecificClassesSpecification";
import { RelatedInstanceNodesSpecification } from "./RelatedInstanceNodesSpecification";
import { CustomQueryInstanceNodesSpecification } from "./CustomQueryInstanceNodesSpecification";
import { RelatedInstanceSpecification } from "../RelatedInstanceSpecification";

/**
 * Used for serializing array of [[ChildNodeSpecification]]
 * @public
 */
export enum ChildNodeSpecificationTypes {
  // hierarchy specifications
  AllInstanceNodes = "AllInstanceNodes",
  AllRelatedInstanceNodes = "AllRelatedInstanceNodes",
  RelatedInstanceNodes = "RelatedInstanceNodes",
  InstanceNodesOfSpecificClasses = "InstanceNodesOfSpecificClasses",
  CustomQueryInstanceNodes = "CustomQueryInstanceNodes",
  CustomNode = "CustomNode",
}

/**
 * Base interface for all [[ChildNodeSpecification]] implementations. Not
 * meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface ChildNodeSpecificationBase {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes;

  /**
   * Defines the order in which specifications are evaluated and executed. Defaults to `1000`.
   *
   * @type integer
   */
  priority?: number;

  /**
   * This tells the rules engine that nodes produced using this
   * specification always or never have children. Defaults to `Unknown`.
   *
   * **Note:** setting this flag to `Always` or `Never` improves performance.
   */
  hasChildren?: "Always" | "Never" | "Unknown";

  /**
   * Hide nodes provided by this specification and directly show nodes of its children.
   * This helps if you need to define related instance nodes of particular parent node that is not available in the
   * hierarchy.
   */
  hideNodesInHierarchy?: boolean;

  /**
   * Hide nodes if they don't have children.
   */
  hideIfNoChildren?: boolean;

  /**
   * An [ECExpression]($docs/learning/presentation/Hierarchies/ECExpressions.md#specification) which
   * indicates whether a node should be hidden or not.
   *
   * @note While the attribute provides much flexibility, it also has performance implications - it's
   * strongly suggested to first consider using `instanceFilter`, `hideNodesInHierarchy` or `hideIfNoChildren`
   * and only use `hideExpression` if none of them are sufficient.
   *
   * @beta
   */
  hideExpression?: string;

  /**
   * Set this flag to `true` to suppress default sorting of ECInstances returned by this specification.
   *
   * **Note:** setting this flag to `true` improves performance.
   */
  doNotSort?: boolean;

  /** Specifications of related instances that can be used in content creation. */
  relatedInstances?: RelatedInstanceSpecification[];

  /** [Nested rule]($docs/learning/presentation/Hierarchies/Terminology.md#nested-rules) specifications. */
  nestedRules?: ChildNodeRule[];
}

/**
 * A container of default grouping properties. Used for specifications that support
 * default grouping. Not meant to be used directly, see `ChildNodeSpecification`.
 *
 * @public
 */
export interface DefaultGroupingPropertiesContainer {
  /** Group instances by ECClass. Defaults to `true`. */
  groupByClass?: boolean;

  /** Group instances by label. Defaults to `true`. */
  groupByLabel?: boolean;
}

/**
 * Navigation rule specifications that define what content the rule results in.
 * @public
 */
export type ChildNodeSpecification = AllInstanceNodesSpecification |
  AllRelatedInstanceNodesSpecification |
  CustomNodeSpecification |
  InstanceNodesOfSpecificClassesSpecification |
  RelatedInstanceNodesSpecification |
  CustomQueryInstanceNodesSpecification;
