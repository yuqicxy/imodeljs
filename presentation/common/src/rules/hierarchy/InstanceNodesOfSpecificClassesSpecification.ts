/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { ChildNodeSpecificationBase, ChildNodeSpecificationTypes, DefaultGroupingPropertiesContainer } from "./ChildNodeSpecification";
import { MultiSchemaClassesSpecification } from "../ClassSpecifications";

/**
 * Creates nodes for all instances of specified ECClasses.
 * @public
 */
export interface InstanceNodesOfSpecificClassesSpecification extends ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer {
  /** Used for serializing to JSON. */
  specType: ChildNodeSpecificationTypes.InstanceNodesOfSpecificClasses;

  /**
   * Specifications of ECClasses whose instances should be returned.
   */
  classes: MultiSchemaClassesSpecification | MultiSchemaClassesSpecification[];

  /**
   * Should all [[classes]] be handled polymorphically.
   */
  arePolymorphic?: boolean;

  /**
   * Condition for filtering instances of defined classes.
   *
   * **See:** [ECExpressions Available in InstanceFilter]($docs/learning/presentation/Hierarchies/ECExpressions.md#instance-filter).
   */
  instanceFilter?: string;
}
