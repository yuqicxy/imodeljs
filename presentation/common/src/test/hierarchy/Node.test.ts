/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import { Node, NodeJSON } from "../../hierarchy/Node";
import {
  createRandomECInstanceNode, createRandomECInstanceNodeKey,
  createRandomECInstanceKeyJSON,
  createRandomLabelDefinitionJSON,
} from "../_helpers/random";
import { ECInstanceNodeKeyJSON } from "../../hierarchy/Key";

const createRandomECInstanceNodeKeyJSON = (): ECInstanceNodeKeyJSON => {
  return {
    ...createRandomECInstanceNodeKey(),
    instanceKey: createRandomECInstanceKeyJSON(),
  };
};

const createRandomNodeJSON = (): NodeJSON => {
  return {
    ...createRandomECInstanceNode(),
    key: createRandomECInstanceNodeKeyJSON(),
    labelDefinition: createRandomLabelDefinitionJSON(),
  };
};

describe("Node", () => {

  describe("toJSON", () => {

    it("serializes Node", () => {
      const node = createRandomECInstanceNode();
      const json = Node.toJSON(node);
      expect(json).to.matchSnapshot();
    });

    it("serializes Node without labelDefinition", () => {
      const node = createRandomECInstanceNode();
      node.labelDefinition = undefined;
      const json = Node.toJSON(node);
      expect(json).to.matchSnapshot();
    });

  });

  describe("fromJSON", () => {

    it("creates valid Node from JSON", () => {
      const json = createRandomNodeJSON();
      const node = Node.fromJSON(json);
      expect(node).to.matchSnapshot();
    });

    it("creates valid Node from serialized JSON", () => {
      const json = createRandomNodeJSON();
      const node = Node.fromJSON(JSON.stringify(json));
      expect(node).to.matchSnapshot();
    });

  });

  describe("listFromJSON", () => {

    it("creates valid Node[] from JSON", () => {
      const json = [createRandomNodeJSON(), createRandomNodeJSON()];
      const nodes = Node.listFromJSON(json);
      expect(nodes).to.matchSnapshot();
    });

    it("creates valid Node[] from serialized JSON", () => {
      const json = [createRandomNodeJSON(), createRandomNodeJSON()];
      const nodes = Node.listFromJSON(JSON.stringify(json));
      expect(nodes).to.matchSnapshot();
    });

  });

});
