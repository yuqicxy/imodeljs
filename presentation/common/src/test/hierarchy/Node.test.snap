// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node fromJSON creates valid Node from JSON 1`] = `
Object {
  "backColor": "rgb(
    182,
    131,
    203
  )",
  "description": "Ab ea eveniet ullam occaecati aperiam non ab.",
  "foreColor": undefined,
  "hasChildren": false,
  "imageId": "bandwidth",
  "isCheckboxEnabled": true,
  "isCheckboxVisible": false,
  "isChecked": true,
  "isEditable": false,
  "isExpanded": false,
  "isSelectionDisabled": true,
  "key": Object {
    "instanceKey": Object {
      "className": "Afghanistan",
      "id": "0x13ce90000002ea7",
    },
    "pathFromRoot": Array [
      "5129f82b-5276-4d87-8ba3-b8ff74048811",
      "93c08eb4-5a77-4ddb-ba30-f33c97571e0f",
    ],
    "type": "ECInstanceNode",
  },
  "label": "utilize",
  "labelDefinition": Object {
    "displayValue": "utilize",
    "rawValue": "Central",
    "typeName": "string",
  },
}
`;

exports[`Node fromJSON creates valid Node from serialized JSON 1`] = `
Object {
  "foreColor": "#D7FE8A",
  "hasChildren": false,
  "imageId": "Credit Card Account",
  "isCheckboxEnabled": false,
  "isCheckboxVisible": false,
  "isChecked": true,
  "isEditable": false,
  "isExpanded": false,
  "isSelectionDisabled": false,
  "key": Object {
    "instanceKey": Object {
      "className": "24/7",
      "id": "0x13072000000c8dc",
    },
    "pathFromRoot": Array [
      "41ce5242-abda-42a7-96dd-f8d9fe71dab8",
      "e4358d78-e06d-4c62-8f5a-02b9ecb87afe",
    ],
    "type": "ECInstanceNode",
  },
  "label": "website",
  "labelDefinition": Object {
    "displayValue": "website",
    "rawValue": "Forest",
    "typeName": "string",
  },
}
`;

exports[`Node listFromJSON creates valid Node[] from JSON 1`] = `
Array [
  Object {
    "backColor": "rgb(
    114,
    183,
    43
  )",
    "description": undefined,
    "foreColor": undefined,
    "hasChildren": false,
    "imageId": undefined,
    "isCheckboxEnabled": true,
    "isCheckboxVisible": true,
    "isChecked": false,
    "isEditable": false,
    "isExpanded": false,
    "isSelectionDisabled": false,
    "key": Object {
      "instanceKey": Object {
        "className": "Refined Steel Gloves",
        "id": "0x17ac2000000a3a1",
      },
      "pathFromRoot": Array [
        "d20e88ab-2d4b-4a4e-b157-e9efc4ab5ccd",
        "51f65688-18bb-4596-9c06-98a07d760acd",
      ],
      "type": "ECInstanceNode",
    },
    "label": "optical",
    "labelDefinition": Object {
      "displayValue": "optical",
      "rawValue": "Turkmenistan",
      "typeName": "string",
    },
  },
  Object {
    "backColor": "rgb(
    192,
    167,
    181
  )",
    "description": "Earum placeat ut est mollitia soluta.",
    "foreColor": undefined,
    "hasChildren": true,
    "imageId": undefined,
    "isCheckboxEnabled": true,
    "isCheckboxVisible": true,
    "isChecked": true,
    "isEditable": true,
    "isExpanded": false,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "Plastic",
        "id": "0x4b3f0000000e4e",
      },
      "pathFromRoot": Array [
        "3e690a05-18d9-42dd-989f-9a8a212f6872",
        "fd8f9f65-9cef-49c2-9bb4-4acc219c15db",
      ],
      "type": "ECInstanceNode",
    },
    "label": "sensor",
    "labelDefinition": Object {
      "displayValue": "sensor",
      "rawValue": "Money Market Account",
      "typeName": "string",
    },
  },
]
`;

exports[`Node listFromJSON creates valid Node[] from serialized JSON 1`] = `
Array [
  Object {
    "hasChildren": true,
    "imageId": "Savings Account",
    "isCheckboxEnabled": false,
    "isCheckboxVisible": true,
    "isChecked": true,
    "isEditable": false,
    "isExpanded": false,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "Borders",
        "id": "0xb50a0000000fa4",
      },
      "pathFromRoot": Array [
        "f2825eae-2829-48de-ba99-b791d53bd36d",
        "463293bd-6f88-4836-ba7b-bcc6116044ce",
      ],
      "type": "ECInstanceNode",
    },
    "label": "circuit",
    "labelDefinition": Object {
      "displayValue": "circuit",
      "rawValue": "bluetooth",
      "typeName": "string",
    },
  },
  Object {
    "foreColor": "#903169",
    "hasChildren": false,
    "isCheckboxEnabled": false,
    "isCheckboxVisible": true,
    "isChecked": true,
    "isEditable": false,
    "isExpanded": false,
    "isSelectionDisabled": true,
    "key": Object {
      "instanceKey": Object {
        "className": "Libyan Dinar",
        "id": "0x1118a000000961c",
      },
      "pathFromRoot": Array [
        "0047b71e-154a-47dd-9bf1-d5a969ce80cc",
        "7a46b3f4-ba8b-41b8-b5df-1f018740c508",
      ],
      "type": "ECInstanceNode",
    },
    "label": "Indiana",
    "labelDefinition": Object {
      "displayValue": "Indiana",
      "rawValue": "web-enabled",
      "typeName": "string",
    },
  },
]
`;

exports[`Node toJSON serializes Node 1`] = `
Object {
  "backColor": undefined,
  "description": "Est esse quo dicta ut.",
  "foreColor": "#553B65",
  "hasChildren": false,
  "imageId": undefined,
  "isCheckboxEnabled": true,
  "isCheckboxVisible": false,
  "isChecked": true,
  "isEditable": true,
  "isExpanded": true,
  "isSelectionDisabled": true,
  "key": Object {
    "instanceKey": Object {
      "className": "International",
      "id": "0x15d4700000161fd",
    },
    "pathFromRoot": Array [
      "1a565950-ae3f-4e3b-839c-1d6f2a17055c",
      "accb47fb-8ce5-438e-885a-20877ad8fc1f",
    ],
    "type": "ECInstanceNode",
  },
  "label": "methodical",
  "labelDefinition": Object {
    "displayValue": "methodical",
    "rawValue": "Berkshire",
    "typeName": "string",
  },
}
`;

exports[`Node toJSON serializes Node without labelDefinition 1`] = `
Object {
  "backColor": "rgb(
    153,
    118,
    239
  )",
  "description": undefined,
  "foreColor": undefined,
  "hasChildren": false,
  "imageId": "parsing",
  "isCheckboxEnabled": true,
  "isCheckboxVisible": true,
  "isChecked": false,
  "isEditable": true,
  "isExpanded": false,
  "isSelectionDisabled": false,
  "key": Object {
    "instanceKey": Object {
      "className": "National",
      "id": "0x1578700000143e5",
    },
    "pathFromRoot": Array [
      "ea188cf7-d7fb-441a-ac80-5d650de79565",
      "9ee961c1-df2f-46a2-8c6e-57e6f9c31ed3",
    ],
    "type": "ECInstanceNode",
  },
  "label": "transmit",
  "labelDefinition": Object {
    "displayValue": "transmit",
    "rawValue": "transmit",
    "typeName": "string",
  },
}
`;
